lib/collection/src/shards/replica_set/mod.rs

```diff
@@ -523,29 +526,34 @@ impl ShardReplicaSet {
     /// # Cancel safety
     ///
     /// This method is cancel safe.
-    async fn wait_for<F>(&self, check: F, timeout: Duration) -> CollectionResult<()>
+    pub fn wait_for<F>(
+        &self,
+        check: F,
+        timeout: Duration,
+    ) -> impl Future<Output = CollectionResult<()>> + 'static
     where
         F: Fn(&ReplicaSetState) -> bool + Send + 'static,
     {
         // TODO: Propagate cancellation into `spawn_blocking` task!?
 
         let replica_state = self.replica_state.clone();
-        let timed_out =
-            !tokio::task::spawn_blocking(move || replica_state.wait_for(check, timeout))
-                .await
-                .map_err(|err| {
-                    CollectionError::service_error(format!(
-                        "Failed to wait for replica set state: {err}"
-                    ))
-                })?;
-
-        if timed_out {
-            return Err(CollectionError::service_error(
-                "Failed to wait for replica set state, timed out",
-            ));
-        }
+        let task = tokio::task::spawn_blocking(move || replica_state.wait_for(check, timeout));
 
-        Ok(())
+        async move {
+            let status = task.await.map_err(|err| {
+                CollectionError::service_error(format!(
+                    "Failed to wait for replica set state: {err}"
+                ))
+            })?;
+
+            if status {
+                Ok(())
+            } else {
+                Err(CollectionError::service_error(
```

Comment:
I checked usages, it should be safe to convert this error to timeout. Both Service Error and timeout are transient, also it looks like all actual places where it is used either don't care or force convert error into service error.